changes to unaligned_jmp tests to support unaligned support

allenjbaum opened the pull request:
Changed tests to support both unaligned support (C-ext) and without. 
Added lots of comments, changed signature to be fixed length, 
   first word test#, next 3 are trap parameters if it traps, else default value 
Added one test for unaligned JAL that shouldn't trap 
Removed attempt to clear MISA.C bit Added a bunch of FIXMEs

neelgala commented:
I've been looking at the changes and would suggest the following:

1. line 367 should be `csrw  mepc, x30` . Get rid of the extra 'r'
2. You are using explicit compressed instructions: ` c.li and c.nop`. So you will have to move them to the rv32imc so that the test gets compiled with the `-march=rv32imc`. 
3. You will have to ensure that you use `.option rvc and .option norvc` to make sure not everything gets compressed. 
4. Also make sure you don't declare too much signature section for a part. For eg, test-part-A1 performs a single store to the signature but you have allocated `.fill 4, 4, -1`. It should be `.fill 1, 4, -1`. Probably ensure the same for all other parts. 
5. I can compile the test with `march=rv32imc` and it probably will execute on spike. But since you are using `c.li and c.nop` explicitly compiling without 'c' march field won't work. And that is what you need to test the second part of your claim - that the test will work without un-aligned support. I would rather suggest having 2 different tests one which supports un-aligned and another that does not, since you can't compile them for the same target.

allenjbaum commented:
A few points:
1. I'll fix that
2, 3. I deliberately wanted to add the c.li and c.nop, but if the assembler
can't ahandler expicitly mixed code, I'll rewrite.
   I do need to deliberately place unaligned instructions to perform the
tests; I'm not quite sure how to do that without having an explicit
compressed op (I can easily align, but how to unalign? Fill in with data?)
4. I'll look again, but I think that was also deliberate, to ensure that
the signature was fixed length regardless of the path it took (trap or not
trap)
5. Do you know if the .rvc, .norvc options can be used dynamically, e.g.
switching back and forth  in the middle of the code?
 That would fix most of the issues you've noted.

On Thu, May 23, 2019 at 7:33 AM Neel Gala <notifications@github.com> wrote:

> I've been looking at the changes and would suggest the following:
>
>    1. line 367 should be csrw mepc, x30 . Get rid of the extra 'r'
>    2. You are using explicit compressed instructions: c.li and c.nop. So
>    you will have to move them to the rv32imc so that the test gets compiled
>    with the -march=rv32imc.
>    3. You will have to ensure that you use .option rvc and .option norvc
>    to make sure not everything gets compressed.
>    4. Also make sure you don't declare too much signature section for a
>    part. For eg, test-part-A1 performs a single store to the signature but you
>    have allocated .fill 4, 4, -1. It should be .fill 1, 4, -1. Probably
>    ensure the same for all other parts.
>    5. I can compile the test with march=rv32imc and it probably will
>    execute on spike. But since you are using c.li and c.nop explicitly
>    compiling without 'c' march field won't work. And that is what you need to
>    test the second part of your claim - that the test will work without
>    un-aligned support. I would rather suggest having 2 different tests one
>    which supports un-aligned and another that does not, since you can't
>    compile them for the same target.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/riscv-compliance/pull/55?email_source=notifications&email_token=AHPXVJV3NMHJTOPZOWXDAK3PW2TMPA5CNFSM4HDODZFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWCNMEY#issuecomment-495244819>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHPXVJSE7BR6FS6KYF3GLJLPW2TMPANCNFSM4HDODZFA>
> .
>

neelgala commented:
Yes you can use `.option rvc` or `.option norvc` within the code. You could probably also subsume them under a macro defined at compile time as well.

jstraus59 commented:
Allen
I think there needs to be some discussion on some of these issues before we
transform things too much - for example change signature to be fixed length
- have you gone and changed all the tests, signatures etc in the repo or
with the document be all about futures that will all need discussing?
thanks
Simon


On Wed, Apr 3, 2019 at 11:03 PM allenjbaum <notifications@github.com> wrote:

> Changed tests to support both unaligned support (C-ext) and without.
> Added lots of comments, changed signature to be fixed length,
> first word test#, next 3 are trap parameters if it traps, else default
> value
> Added one test for unaligned JAL that shouldn't trap
> Removed attempt to clear MISA.C bit Added a bunch of FIXMEs
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/riscv/riscv-compliance/pull/55
> Commit Summary
>
>    - iadded details, minor corrections, some ToBeDiscussed items
>    - Changed tests to support both unaligned support (C-ext) and without.
>
> File Changes
>
>    - *M* riscv-test-suite/rv32i/src/I-MISALIGN_JMP-01.S
>    <https://github.com/riscv/riscv-compliance/pull/55/files#diff-0> (238)
>    - *M* spec/TestFormatSpec.adoc
>    <https://github.com/riscv/riscv-compliance/pull/55/files#diff-1> (47)
>
> Patch Links:
>
>    - https://github.com/riscv/riscv-compliance/pull/55.patch
>    - https://github.com/riscv/riscv-compliance/pull/55.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/riscv-compliance/pull/55>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJrER6XHhrjPOcIb2_qgTayNwTlesnvXks5vdST7gaJpZM4cbh5K>
> .
>


-- 
====================================================================
The information contained in this electronic mail message and any
attachments hereto
is privileged and confidential information intended only for the use of the
individual or
entity named above or their designee.  If the reader of this message is not
the intended
recipient, you are hereby notified that any dissemination, distribution or
copying of this
communication is strictly prohibited.  If you have received this
communication in error
please immediately notify us by return  message or by telephone and delete
the
original message from your mail system.  Thank you.
====================================================================

allenjbaum commented:
This is specifically for the misaligned test (the only one I'm updating),
and specifically to fix an issue with tests that have multiple parts and
can trap (or not, depending on core configuration).
I am currently modifying the test to expand coverage and fix issues
identified by Neel.

On Wed, May 29, 2019 at 7:26 AM Jim Straus <notifications@github.com> wrote:

> Allen
> I think there needs to be some discussion on some of these issues before we
> transform things too much - for example change signature to be fixed length
> - have you gone and changed all the tests, signatures etc in the repo or
> with the document be all about futures that will all need discussing?
> thanks
> Simon
>
>
> On Wed, Apr 3, 2019 at 11:03 PM allenjbaum <notifications@github.com>
> wrote:
>
> > Changed tests to support both unaligned support (C-ext) and without.
> > Added lots of comments, changed signature to be fixed length,
> > first word test#, next 3 are trap parameters if it traps, else default
> > value
> > Added one test for unaligned JAL that shouldn't trap
> > Removed attempt to clear MISA.C bit Added a bunch of FIXMEs
> > ------------------------------
> > You can view, comment on, or merge this pull request online at:
> >
> > https://github.com/riscv/riscv-compliance/pull/55
> > Commit Summary
> >
> > - iadded details, minor corrections, some ToBeDiscussed items
> > - Changed tests to support both unaligned support (C-ext) and without.
> >
> > File Changes
> >
> > - *M* riscv-test-suite/rv32i/src/I-MISALIGN_JMP-01.S
> > <https://github.com/riscv/riscv-compliance/pull/55/files#diff-0> (238)
> > - *M* spec/TestFormatSpec.adoc
> > <https://github.com/riscv/riscv-compliance/pull/55/files#diff-1> (47)
> >
> > Patch Links:
> >
> > - https://github.com/riscv/riscv-compliance/pull/55.patch
> > - https://github.com/riscv/riscv-compliance/pull/55.diff
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/riscv/riscv-compliance/pull/55>, or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AJrER6XHhrjPOcIb2_qgTayNwTlesnvXks5vdST7gaJpZM4cbh5K
> >
> > .
> >
>
>
> --
> ====================================================================
> The information contained in this electronic mail message and any
> attachments hereto
> is privileged and confidential information intended only for the use of the
> individual or
> entity named above or their designee. If the reader of this message is not
> the intended
> recipient, you are hereby notified that any dissemination, distribution or
> copying of this
> communication is strictly prohibited. If you have received this
> communication in error
> please immediately notify us by return message or by telephone and delete
> the
> original message from your mail system. Thank you.
> ====================================================================
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/riscv-compliance/pull/55?email_source=notifications&email_token=AHPXVJQKN4P45LKZY52FDV3PX2HCJA5CNFSM4HDODZFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWPQFZI#issuecomment-496960229>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHPXVJSRY3EWPPYWND7YQNLPX2HCJANCNFSM4HDODZFA>
> .
>

allenjbaum commented:
I've totally updated the test, . There are a LOT of changes, primarily
comments, bit there were also some bugs i found by inspection. I've also
improved the trap handler (and realized that the framework will need to
deal with telling the difference between a RV_COMPLIANCE_HALT and any other
kind of halt (I save the current trap vector in X31. Inside the handler I
save/restore X31 into mscratch - that was a bug before. The trap handler
saves 3 words as signature, but the test without trapping only saves one,
so  I always bump the counter by 3 words to ensure signature checking
doesn't get messed up.

The test spec I'm updating is attached. Let me know if you see any bugs or
objections. Next up: reviewing you WARL doc.

On Thu, May 23, 2019 at 7:33 AM Neel Gala <notifications@github.com> wrote:

> I've been looking at the changes and would suggest the following:
>
>    1. line 367 should be csrw mepc, x30 . Get rid of the extra 'r'
>    2. You are using explicit compressed instructions: c.li and c.nop. So
>    you will have to move them to the rv32imc so that the test gets compiled
>    with the -march=rv32imc.
>    3. You will have to ensure that you use .option rvc and .option norvc
>    to make sure not everything gets compressed.
>    4. Also make sure you don't declare too much signature section for a
>    part. For eg, test-part-A1 performs a single store to the signature but you
>    have allocated .fill 4, 4, -1. It should be .fill 1, 4, -1. Probably
>    ensure the same for all other parts.
>    5. I can compile the test with march=rv32imc and it probably will
>    execute on spike. But since you are using c.li and c.nop explicitly
>    compiling without 'c' march field won't work. And that is what you need to
>    test the second part of your claim - that the test will work without
>    un-aligned support. I would rather suggest having 2 different tests one
>    which supports un-aligned and another that does not, since you can't
>    compile them for the same target.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/riscv-compliance/pull/55?email_source=notifications&email_token=AHPXVJV3NMHJTOPZOWXDAK3PW2TMPA5CNFSM4HDODZFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWCNMEY#issuecomment-495244819>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHPXVJSE7BR6FS6KYF3GLJLPW2TMPANCNFSM4HDODZFA>
> .
>

neelgala commented:
Will go through it over the weekend and update early next week.

eroom1966 commented:
I note this is still pending and requires conflicts to he resolved ?

jeremybennett commented:
The branch needs to be rebased on master and the conflicts in the following files resolved
- `ChangeLog`
- `riscv-test-suite/rv32i/src/I-MISALIGN_JMP-01.S`
- `spec/TestFormatSpec.adoc`

Then force push your branch and it will be mergeable without conflicts.
